### DISCUSSION TITLE
make usage of "-fpermissive" configuarable

### DISCUSSION THREAD

**Discussion by Lorandil** [SpenceKonde/ATTinyCore#639] [Development Plans]
Hi all,

lately I had trouble with two kinds of bugs:
* declaring variables in switch statement without block 
* a return path of a "uint8_t" function not returning a value

I would have expected these coding errors to be reported as error during compile time,
instead of resulting unpredictable behaviour. 

My understanding is, that this is because of the "-fpermissive" directive which is enabled by default in the 
ATtiny core (and the Arduino AVR core).

I would really appreciate if there would be a possibility to disable this "feature", I cannot see any reason for its existence
except of compatibility to some old code (which should be rewritten!)...

**Comment by SpenceKonde** [comment]
Pretty sure those should be reported as warnings (as are all the other cases where -fpermissive allows something). However, you probably have warnings disabled because the Arduino IDE defaults to turning off all warnings. On my other cores I recently got fed up with that behavior (partly after I was surprised to find that one of the dozen or so copies of the IDE I had installed and was using for testing didn't have warnings enabled) and made platform.txt  enable warnings regardless of the IDE's configuration, and ATTinyCore is getting that treatment in 2.0.0 - it's just not acceptable to have the *default* be to disable that critical source of debugging information. I can't change the IDE defaults though.

With that treatment, I don't think there should be a need to make -fpermissive a menu option (the reason we have that option is that a huge amount of code in the wild breaks without it, generally written by people who are unaware that they're relying on it, even though they're writing widely used libraries. Because the IDE disables warnings....

The vast majority of users don't realize that a) warnings can be turned off or on in preferences screen, b) that they're turned off by default, and/or c) that warnings are rarely benign. and I decided it was better to annoy the few people who both know about the option and wish to leave it off than to let the IDE default to wasting the time of everyone else.

**Reply by Lorandil** [reply]
Thanks for the clarification.
I must admit that I already had warnings turned on,
I simply didn't see the warning lines (I'm using IDE 2.0, where warnings and errors aren't highlighted anymore as in 1.8.xx... I should write those guys later ;)). 
I'm still shocked that broken code doesn't cause a compiler error by default, but I understand that is an Arduino IDE "requirement".
